libbeat/processors/rate_limit/algorithm/token_bucket.go

```diff
@@ -0,0 +1,114 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package algorithm
+
+import (
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+func init() {
+	register("token_bucket", newTokenBucket)
+}
+
+type bucket struct {
+	tokens        float64
+	lastReplenish time.Time
+}
+
+type tokenBucket struct {
+	limit   Rate
+	depth   float64
+	buckets map[string]bucket
+}
+
+func newTokenBucket(config Config) (Algorithm, error) {
+	var cfg struct {
+		BurstMultiplier float64 `config:"burst_multiplier"`
+	}
+
+	if err := config.Config.Unpack(&cfg); err != nil {
+		return nil, errors.Wrap(err, "could not unpack token_bucket algorithm configuration")
+	}
+
+	return &tokenBucket{
+		config.Limit,
+		config.Limit.value * cfg.BurstMultiplier,
+		make(map[string]bucket, 0),
+	}, nil
+}
+
+func (t *tokenBucket) IsAllowed(key string) bool {
+	t.replenishBuckets()
```

Comment:
I like your GC approach and will implement it. Should we expose (but not necessarily document) options for controlling the GC behavior? I have a feeling choosing the thresholds for when to GC is going to be somewhat arbitrary so providing some knobs to tune it might be useful, in case users start experiencing memory issues after release.

Reply:
Implemented in 0d285ef1328add8479bf3b365c8815f8f3a60d11.